### PR TITLE
Set and Map API improvements

### LINKED ISSUE #11334: Add Map.add_to_{list,set}
Here are two functions I keep on defining after a `Map.Make`: 

```ocaml 
val Map.add_to_list : Map.key -> 'a -> 'a list Map.t -> 'a list Map.t
(** [Map.add_to_list k v m] is [m] with [k] mapping to [l] such that [l] is
    [v :: Map.find k m] if [k] was bound in [m] and [[v]] otherwise. *)

val Map.add_to_set :
      (module Set.S with type elt = 'a and type t = 'set) ->
      Map.key -> 'a -> 'set Map.t -> 'set Map.t
(** [Map.add_to_set (module S) k v m] is [m] with [k] mapping to [s] such that [s] is
    [S.add v (Map.find k m)] if [k] was bound in [m] and [S.singleton v] otherwise. *)
```

I'd like to add them to `Map` but I prefer to test the waters and bikeshed before trying to PR. Here is a [typical implementation](https://github.com/b0-system/b0/blob/914aeaae0861b185377267ccf1366d030453ae8f/src/b00/std/b00_std.ml#L1222-L1231) of these functions. 

The second function would make `Map` depend on `Set`'s interface which people may understandably object.

### PR DESCRIPTION
**Pull request by dbuenzli**
This PR improves the usability of `Set.S` and `Map.S` as follows:

1. Add documentation sections in `{Set,Map}.S`. The number of functions have grown significantly and it's good to have a few signposts for orientation. Section names follow those that can already be found in `Result`, `Option`, `String` and `List`. See [`Set.S`](https://erratique.ch/tmp/setmap/stdlib/Set/module-type-S/index.html) and [`Map.S`](https://erratique.ch/tmp/setmap/stdlib/Map/module-type-S/index.html) for a rendering.
  
2. Uniformize conversions with list values so that we get `{Map,Set}.{to,of}_list`. This means a) Adding `Set.to_list` as an alias of `Set.elements` b) Adding `Map.to_list` as an alias of `Map.bindings` c) Adding the new function `Map.of_list`.
  
3. Add the function `Map.add_to_list` that was suggested in #11334
   
Since the new labelled/unabelled machinery makes all that quite unconvenient and noisy. Generation steps are made as separate commits and the PR should be reviewed commit by commit. This also entails repo states that do not compile. The person merging this may want to squash everything into a single commit. 

Closes #11334.

### CODE REVIEW

**Comment by c-cube** [comment]
I think that's very useful. Have you considered also having `add_list` to both, where `of_list l = add_list l empty`? It's convenient to add many elements without building a separate map/set and calling `union` (which is less efficient).

**Comment by dbuenzli** [comment]
I haven't. In fact I was not there for all that :-) I'm fine with using the other data structure's fold in the case you mention. For the list conversion stuff I was just trying to uniformise a bit conversions with lists, which was rather unsatisfying in my opinion (`Set.{of_list,elements}` and `Map.bindings` before this PR).

File: stdlib/set.mli
Comment (avsm) [inline stdlib/set.mli:117]: Is odoc smart enough to understand `{!Ord.compare}` these days? That might be a useful cross-reference here.

Comment (dbuenzli) [reply stdlib/set.mli:117]: Likely but it's not the goal of this PR to improve the docstrings of the module.

**Comment by dbuenzli** [comment]
For upstream devs fearing to review this PR, commit-by-commit it's easier than it looks :-) 

I'm willing to remove https://github.com/ocaml/ocaml/pull/11410/commits/c05b3d65a9a7a6584168ce8b21fa21347925fabc if the introduction of dictionary passing (and the `Map` on `Set` dependency it introduces) is the reason of the reluctance. 

These kinds of PRs are a bit a pain to assemble and I'd hate if it would start to conflict.

File: stdlib/templates/set.template.mli
Comment (nojb) [inline stdlib/templates/set.template.mli]:
```suggestion
        @since 5.1.0 *)
```

**Review by nojb — COMMENTED** [review]
I suggest to split the addition of `add_to_set` to be discussed in a separate PR.

Otherwise, LGTM (modulo `add_to_set`). As this PR touches the standard library, a second review is needed.

Thanks!

**Comment by nojb** [comment]
(Note that some test results that contain the signature of the `Set` and `Map` modules need to be updated.)

**Review by nojb — APPROVED** [review]
LGTM

**Review by gasche — APPROVED** [review]
This also looks good to me. (I wasn´t brave enough to suggest splitting the new function, but I agree that it makes it much simpler to come to a decision.)

**Comment by dbuenzli** [comment]
Thank you all for the reviews and approvals. 

> (I wasn´t brave enough to suggest splitting the new function, but I agree that it makes it much simpler to come to a decision.)

Well it was a badly packaged PR. For those who are interested in the fate of the removed function (`Map.add_to_set`) I made my autocriticism [here](https://github.com/ocaml/ocaml/issues/11334#issuecomment-1263733173).

**Pull request merged**
